### PR TITLE
Support for iframes

### PR DESCRIPTION
**Pull request by grouma**
- An iframe will have a corresponding evaluation context
- Create an abstraction `EvaluationContext` to track this context
- Update all `Runtime.evaluate` calls to use the tracked context
- Update connection logic to search through all contexts in a given tab to find the Dart application

### CODE REVIEW

File: dwds/debug_extension/web/background.dart [hunk 1]
```diff
@@ -30,31 +30,64 @@ void main() {
...
+            contexts.add(context['id'] as int);
+          }
         }));
+
+        sendCommand(Debuggee(tabId: currentTab.id), 'Runtime.enable',
+            EmptyParam(), allowInterop((e) {}));
+
+        // There is no way to calculate the number of existing execution contexts
+        // so we wait for a short while to recieve events.
+        await Future.delayed(const Duration(milliseconds: 50));
```
Comment (jakemac53) [inline dwds/debug_extension/web/background.dart]: This needs to be something more robust. Why not a stream of contexts or something like that instead?

Comment (jakemac53) [reply dwds/debug_extension/web/background.dart]: We could always still add a timeout still but that would allow it to be a much longer timeout without hampering performance in the good case.

Comment (alan-knight) [reply dwds/debug_extension/web/background.dart]: Do execution contexts map directly to iframes? And what does 'existing' mean? Before a reload? We can get the iframe tree, so if they directly correspond that would work. Or is it that they might not be created yet, or there might be a different number after reload?

Comment (grouma) [reply dwds/debug_extension/web/background.dart]: Execution contexts do not map directly to iframes. There will be at least one context for each iframe but it isn't a one to one relationship.

> This needs to be something more robust. Why not a stream of contexts or something like that instead?

@jakemac53, we can have a stream but that doesn't help. In fact that's originally how I had it implemented. The issue is that we don't know how many items to look through on the stream.

> We could always add a timeout but that would allow it to be a much longer timeout without hampering performance in the good case.

I'm not sure what you mean by "good case".

> And what does 'existing' mean?

When you enable the runtime it will send an event for each created context. One of those contexts should contain the Dart metadata.

Comment (jakemac53) [reply dwds/debug_extension/web/background.dart]: > @jakemac53, we can have a stream but that doesn't help. In fact that's originally how I had it implemented. The issue is that we don't know how many items to look through on the stream.

It allows us to handle the contexts as they come in - we can cancel the stream listener as soon as we find a matching one instead of waiting an arbitrary amount of time.

Eventually it can still time out but the delay before we decide to do that can be much longer (and thus be more robust for larger apps etc).

> I'm not sure what you mean by "good case".

The case where we immediately get a matching context

Comment (grouma) [reply dwds/debug_extension/web/background.dart]: I think I follow now. I can change it so that we iterate through contexts, waiting at most 50 ms. That way we won't pay the cost if the first context contains the Dart application.

Comment (jakemac53) [reply dwds/debug_extension/web/background.dart]: We should allow it to take longer than 50ms though - that is really the part that concerns me. It should be fine to wait at least one second if not several seconds to support longer delays here.

Comment (grouma) [reply dwds/debug_extension/web/background.dart]: That will cause issues. If we allow for up to a second delay, it will cause obvious slowness in the initial startup of DevTools. These messages are always sent over the local connection and are received almost immediately. We aren't waiting for new contexts.

File: dwds/debug_extension/web/background.dart [hunk 2]
```diff
@@ -30,31 +30,64 @@ void main() {
...
+                currentTab, dwdsVersion);
+            successCompleter.complete(true);
+          }));
+          if (await successCompleter.future) {
+            success = true;
+            break;
+          }
+        }
+        if (!success) {
+          alert('Unable to launch DevTools. This is not Dart application.');
```
Comment (alan-knight) [inline dwds/debug_extension/web/background.dart]: not _a_ Dart application. (I know, it's been that way for a while, but we should fix it)

File: dwds/debug_extension/web/background.dart [hunk 3]
```diff
@@ -30,31 +30,64 @@ void main() {
...
+        sendCommand(Debuggee(tabId: currentTab.id), 'Runtime.enable',
+            EmptyParam(), allowInterop((e) {}));
+
+        // There is no way to calculate the number of existing execution contexts
+        // so we wait for a short while to recieve events.
+        await Future.delayed(const Duration(milliseconds: 50));
+
+        var success = false;
+        for (var contextId in contexts) {
+          var successCompleter = Completer<bool>();
```
Comment (alan-knight) [inline dwds/debug_extension/web/background.dart]: Maybe we could split part of this thing out into a separate method?

File: dwds/debug_extension/web/background.dart [hunk 4]
```diff
@@ -77,7 +110,7 @@ void main() {
 // Initiates a [DevToolsRequest], handles an [ExtensionRequest],
 // and sends an [ExtensionEvent].
 Future<void> startSseClient(String uri, String appId, String instanceId,
-    Tab currentTab, String dwdsVersion) async {
+    int contextId, Tab currentTab, String dwdsVersion) async {
```
Comment (alan-knight) [inline dwds/debug_extension/web/background.dart]: Six is officially too many positional parameters, and you're adding in the middle. Let's change these to be named parameters.

File: dwds/lib/src/debugging/evaluation_context.dart [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+class EvaluationContext {
+  final RemoteDebugger _remoteDebugger;
+
+  // Contexts that may contain a Dart application.“
+  final _contexts = <int>[];
+
+  int _id;
+
+  /// Returns the context ID that contains the running Dart application.
+  Future<int> get currentId async {
```
Comment (alan-knight) [inline dwds/lib/src/debugging/evaluation_context.dart]: I don't think "current" adds information. How about just "id".

File: dwds/lib/src/debugging/evaluation_context.dart [hunk 6]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:dwds/src/debugging/remote_debugger.dart';
+
+/// The evaluation context in which to do remote evaluations.
+class EvaluationContext {
```
Comment (alan-knight) [inline dwds/lib/src/debugging/evaluation_context.dart]: Chrome calls it ExecutionContext, is there a reason we shouldn't use the same terminology?

File: dwds/lib/src/services/chrome_proxy_service.dart [hunk 7]
```diff
@@ -105,18 +108,19 @@ class ChromeProxyService implements VmServiceInterface {
     AppConnection appConnection,
     LogWriter logWriter,
     bool restoreBreakpoints,
+    EvaluationContext evalContext,
```
Comment (alan-knight) [inline dwds/lib/src/services/chrome_proxy_service.dart]: https://memegen.googleplex.com/5241610888282112

TL;DR - Just 'context'

File: dwds/debug_extension/web/background.dart (hunk 2 above)
Comment (grouma) [reply dwds/debug_extension/web/background.dart]: Done.

File: dwds/debug_extension/web/background.dart (hunk 3 above)
Comment (grouma) [reply dwds/debug_extension/web/background.dart]: Done.

File: dwds/debug_extension/web/background.dart (hunk 4 above)
Comment (grouma) [reply dwds/debug_extension/web/background.dart]: I don't want them named because they are all required and I don't feel like adding another dependency. It's only called in one location so I made it private and the passed arguments are clear given the variable name.

File: dwds/lib/src/debugging/evaluation_context.dart (hunk 5 above)
Comment (grouma) [reply dwds/lib/src/debugging/evaluation_context.dart]: Agreed. Changed.

File: dwds/lib/src/debugging/evaluation_context.dart (hunk 6 above)
Comment (grouma) [reply dwds/lib/src/debugging/evaluation_context.dart]: Changed.

File: dwds/debug_extension/web/background.dart (hunk 1 above)
Comment (grouma) [reply dwds/debug_extension/web/background.dart]: I've updated the solution. We won't pay the cost of the delay if we find a context with the Dart metadata. I don't see any other way without impacting the non-optimal case.
